python/rmm/rmm.py

```diff
@@ -267,6 +267,43 @@ def get_ipc_handle(ary, stream=0):
     )
 
 
+try:
+    import cupy
+except ImportError:
+    cupy = None
+
+if cupy:
+
+    class RMMCuPyMemory(cupy.cuda.memory.BaseMemory):
+        def __init__(self, size):
+            self.size = size
+            if size > 0:
+                self.rmm_array = librmm.device_buffer.DeviceBuffer(size=size)
+                self.ptr = self.rmm_array.ptr
+                self.device_id = cupy.cuda.runtime.pointerGetAttributes(
+                    self.ptr
+                ).device
+            else:
+                self.rmm_array = None
+                self.ptr = 0
+                self.device_id = cupy.cuda.device.get_device_id()
+
+
+def rmm_cupy_allocator(nbytes):
+    """
+    A CuPy allocator that make use of RMM.
+
+    Examples
+    --------
+    >>> import rmm
+    >>> import cupy
+    >>> cupy.cuda.set_allocator(rmm.rmm_cupy_allocator)
+    """
+    import cupy
+
+    return cupy.cuda.memory.MemoryPointer(RMMCuPyMemory(nbytes), 0)
```

Comment:
Where is the memory freed?

Reply:
Admittedly there are a few layers of indirection, but here is the ownership as far as Python is concerned.

`MemoryPointer` -> `RMMCuPyMemory` -> `DeviceBuffer`

`DeviceBuffer` itself hold an `rmm::device_buffer` within a `std::unique_ptr`, which it will free when it scopes out.